pgjdbc/src/main/java/org/postgresql/core/QueryExecutorBase.java

```diff
@@ -41,8 +41,9 @@ public abstract class QueryExecutorBase implements QueryExecutor {
   private final String database;
   private final int cancelSignalTimeout;
 
+  protected int protocol;
   private int cancelPid;
-  private int cancelKey;
+  private byte[] cancelKey = new byte[0];
```

Comment:
Have you considered nullable array instead so we don't allocate the array?

```suggestion
  private byte @Nullable[] cancelKey;
```